packages/extension-commerce-bm-seo/src/setup-app.ts

```diff
@@ -0,0 +1,177 @@
+        const component = Object.assign(() => null, {displayName: `${this.getName()}.${componentName}`})
+
+        return Promise.resolve([
+            {
+                path: requestURL.pathname,
+                component,
+                exact: true
+            }
+        ])
+    }
+
+    /**
+     * TODO: update comment to make it clearer that beforeRouteMatch is also called on the client side
+     * This method is used on the server during the rendering pipeline. It's provided a list of all the routes that your application
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({allRoutes, locals}: BeforeRouteMatchParams): RouteProps[] {
+        const index = allRoutes.findIndex(
+            (route: RouteProps) => 
+                route.component &&
+                route.component.displayName?.includes(this.getName())
+        )
+
+        if (index === -1) {
+            return allRoutes
+        }
+
+        // Complete the partial route
+        const routes = allRoutes.slice()
+        const [route] = routes.splice(index, 1)
+        const componentName = route.component.displayName.split(".").pop()
+
+        if (!componentName) {
+            return allRoutes
+        }
+
+        const component = routes.find((_route: RouteProps) =>
+            _route.component?.displayName?.includes(componentName)
+        )?.component
+
+        if (!component) {
+            // TODO: fix error message
+            throw Error(`Could not find component with displayName "${componentName}"`)
+        }
+
+        route.component = routeComponent(component, true, locals)
+        // NOTE: to be expected: the Sample page will be rendered on the server side, while a different page is then rendered on the client side.
+        // Jinsu's work on serialization will make sure that the same page will be rendered on both server and client sides.
+
+        const result = [route, ...routes]
+        // console.log('--- beforeRouteMatch: resulting routes', result)
+        return result
+    }
+
+    getComponentMap(): ComponentMap {
+        // During deserialize each serialized route in this extension MUST return a component.
+        // This implementation returns an object where the keys are component names in resourceTypeToComponentMap
+        // and the values are placeholder components with the displayName
+        const {resourceTypeToComponentMap} = this.getConfig()
+        return Object.values(resourceTypeToComponentMap).reduce((acc: ComponentMap, name) => {
+            acc[`${this.getName()}.${name}`] = Object.assign(() => null, {displayName: `${this.getName()}.${name}`})
+            return acc
+        }, {} as Record<string, string>)
+    }
```

Comment:
This implementation returns an object where the keys are component names in `resourceTypeToComponentMap` and the values are placeholder components with the `displayName` in the format `CommerceBmSeo.ProductList`